GoogleUtilities/AppDelegateSwizzler/README.md

```diff
@@ -0,0 +1,151 @@
+# App Delegate Swizzler
+
+## Overview
+
+The App Delegate Swizzler swizzles certain methods on the AppDelegate and allows interested parties (for eg. other SDKs like Firebase Analytics) to register listeners when certain App Delegate methods are called. 
+
+The App Delegate Swizzler uses [isa swizzling](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/KeyValueObserving/Articles/KVOImplementation.html) to create a dynamic subclass of the app delegate class and add methods to it that have the logic to add multiple "interceptors".
+
+Adding interceptors to the following methods is currently supported by the App Delegate Swizzler.
+
+* `- (BOOL)application:openURL:options:` [Reference](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1623112-application?language=objc)
+
+Note: This method is added only if the original app delegate implements it. See [the code](https://github.com/firebase/firebase-ios-sdk/blob/a7117ea786f4d8f45b798ceb19b75e5400f2899e/GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m#L313) for a complete explanation.
+
+* `- (BOOL)application:openURL:sourceApplication:annotation:`
+    [Reference](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1623073-application?language=objc)
+
+* `- (void)application:handleEventsForBackgroundURLSession:completionHandler:`
+    [Reference](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1622941-application?language=objc)
+
+* `- (BOOL)application:continueUserActivity:restorationHandler:`
+    [Reference](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1623072-application?language=objc)
+
+We are looking into adding support for more methods as we need them.
+    
+## Adopting the swizzler
+
+To start using the app delegate swizzler to intercept app delegate methods do the following:
+
+The following assumes that you are an SDK that ships using Cocoapods and need to react to one of the app delegate methods listed above.
+
+1. Add a dependency to the app delegate swizzler - `GoogleUtilities/AppDelegateSwizzler:~> 5.2.0`. Note: Pin the version to the minor version because we consider these as internal APIs and can break them with a minor version update. We know this isn't consistent with SemVer and are working to fix it, but for now that's the case.
```

Comment:
We can use semantic versioning for GoogleUtilities, since a major version update does not necessarily imply a major version update for Firebase as long as Firebase's public API's don't do something backwards incompatible.

Reply:
Done.